Extends the I2C slave interface to work on native_posix

sjg20 opened the pull request:
This allows tests of the I2C middle layer and I2C drivers to be written for native_posix.

There is still a limitation of only one 'slave' device per controller. I plan to fix that later.

sjg20 committed:
test: i2c: Document the i2c_slave_api test

This test is not easy to understand. Add a few notes about its
plumbing.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: i2c: Add a dump routine

Add a function to dump out a set of I2C messages. It uses debug logging
so it only useful for debugging.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
tests: i2c: Use variables to improve readability

This test is hard to read because of the use of macros in the body of
the code. The macros differ only in one character. Add variables to
simplify things.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
tests: i2c: Indicate where the failure occurs

At present this tests fails but does not tell you where. Use the
existing assert macros to handle this.

Signed-off-by: Simon Glass <sjg@chromium.org>

aescolar commented on drivers/i2c/Kconfig.i2c_sim:
```diff
@@ -0,0 +1,11 @@
+
+config I2C_SIM
+	bool "I2C simulator"
+	depends on ARCH_POSIX
```
From a quick look at the code: Is this really dependent on the posix arch or native_posix? I mean, can this be used in say QEMU or any other platform with spare RAM? (similarly to the EEPROM_SIMULATOR)

vanwinkeljan replied:
I do agree here that it looks not native posix specific.

sjg20 replied:
Done

aescolar commented on dts/bindings/i2c/zephyr,sim-i2c.yaml:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2020 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+description: Native posix I2C emulator
```
Maybe also update the description to remove the reference to native_posix

Also please in this commit ("drivers: i2c: Implement the slave interface for posix") , change the commit title and description

sjg20 replied:
done

jfischer-phytec-iot commented on drivers/i2c/i2c_sim.c:
```diff
@@ -0,0 +1,128 @@
+	const struct i2c_sim_cfg *config = NULL;
+
+	/* Check for NULL pointers */
+	if (!dev) {
```
This comment is not necessary, you could change it to `if (dev == NULL)`

sjg20 replied:
done

jfischer-phytec-iot commented on drivers/i2c/i2c_sim.c:
```diff
@@ -0,0 +1,128 @@
+	if (!dev) {
+		LOG_ERR("Device handle is NULL");
+		return -EINVAL;
+	}
```
Please add empty lines after if{} blocks

sjg20 replied:
done

tbursztyka commented on drivers/i2c/i2c_common.c:
```diff
@@ -0,0 +1,37 @@
+void i2c_dump_msgs(const char *name, const struct i2c_msg *msgs,
+		   uint8_t num_msgs, uint16_t addr)
+{
+	char data[80], *ptr, *end = data + sizeof(data) - 3;
```
How the '80' has been chosen? (is it enough, are there cases where it would require more?)

sjg20 replied:
Yes if you have > 26 bytes, but that is pretty rate and in those cases you probably don't want to see everything.

I could adjust the code to support it.

sjg20 replied:
Not needed now

pabigot commented on tests/drivers/i2c/i2c_slave_api/src/main.c:
```diff
@@ -142,6 +142,8 @@ void test_eeprom_slave(void)
 	int reg_addr1 = DT_REG_ADDR(DT_INST(1, atmel_at24));
 	const char *label_eeprom0 = DT_LABEL(DT_INST(0, atmel_at24));
```
While this solution works, an approach that would be more realistic with how this should be done is to add distinct node labels to the EEPROM devicetree nodes and use those.  That would eliminate any dependency on which device is assigned which instance.

However, when I do that on the branch immediately before this commit, the test still fails.  Which suggests that the test is wrong.  Which it is.  Multiple calls are using the wrong device address for the I2C controller.

pabigot replied:
This comment is out of date; what it's trying to convey is better described down at the code that tries to figure out which device is which.

pabigot commented on tests/drivers/i2c/i2c_slave_api/doc/index.rst:
```diff
@@ -0,0 +1,55 @@
+.. _i2c_slave_api:
+
```
Standard practice in the tests hierarchy is to put this sort of thing into `README.txt` at the top level of the application directory.  I suppose it could be `README.rst` too, but people aren't going to look for `doc/index.rst` to find this information.

sjg20 replied:
done

pabigot commented on tests/drivers/i2c/i2c_slave_api/src/main.c:
```diff
@@ -202,31 +187,43 @@ void test_eeprom_slave(void)
 
 	LOG_INF("EEPROM %s Attached !", label_eeprom1);
 
+	/* Figure out which slave belongs to which EEPROM */
```
There's something really fishy here.  While the association of an EEPROM instance with a devicetree node can theoretically change between builds, this test is standalone and the devices shouldn't need disambiguation.

When I did this in https://github.com/pabigot/zephyr/commits/pr/26003 the test passes on native_posix (that solution would require adding the node labels to the other overlays as well).  But there's something else funny going on here.  Too much changed in this test for me to figure out what led you to conclude this sort of detection was necessary.  It really shouldn't be.

pabigot replied:
BTW: "this" in the comment above was referencing a paragraph about using the devicetree nodelabel to access a specific EEPROM rather than assuming the instance number was meaningful, which is demonstrated in the commit in my repo.  There was a bunch of commentary about that either in the review summary or the comment that disappeared somehow, including a note that if this path is taken the other devicetree overlays need to be updated to add the node labels.

sjg20 replied:
Each I2C bus has its own AT24 chip. Yes the AT24s use the same driver but they are at different I2C addresses. From my reading I don't think I should rely on the ordering of these devices as referenced by DT_INST() ?

pabigot replied:
That's correct: in a real application you'd be getting them based on the node label (`eeprom_1` and `eeprom_2` in the example I provided).  Doing that ensures you know which one you're getting.

However, in a single program like this where there's no persistance between runs there should be no reason to have to disambiguate the devices: the driver for each doesn't change while the code is running, and the test doesn't care which one was which as long as you consistently use the right address with the right bus controller for each one.

sjg20 replied:
OK. Well I can make it work like it was so I've dropped this one.

The thing that bugged me is that we are setting the contents of each eeprom through one method (a 'direct' call to eeprom_slave_program()) and then reading it via another (using a I2C bus and address). So I wasn't sure that would be a stable mapping. I guess you have confirmed that it is, or that I misunderstand what is going on.

pabigot replied:
As long as the bus and address are obtained from the devicetree node that's used to identify the specific EEPROM yes it's stable.  It's more clear when the node is identified with an explicit name like:

    #define NODE_EP0 DT_INST(0, atmel_at24)
    #define NODE_EP1 DT_INST(1, atmel_at24)

sjg20 replied:
Hmm well I still need to swap it from STM32. So perhaps you can tell me what the solution is...I've pushed the latest code

pabigot replied:
What do you mean you still need to swap it from STM32?  Are you running on hardware with two eeproms?  I see two custom overlays for STM32 boards for the test, one of which mentions an external RF EEPROM connector, though the corresponding daughterboard is apparently obsolete.

How can I reproduce what you see?

sjg20 replied:
I mean that posix needs the devices to be around the other way.

This is what I am using:

 west build -f -b native_posix tests/drivers/i2c/i2c_slave_api/ -d posix 
./posix/zephyr/zephyr.elf

sjg20 commented on drivers/i2c/i2c_common.c:
```diff
@@ -0,0 +1,37 @@
+			*data = '\0';
+		} else {
+			for (j = 0, ptr = data; j < msg->len && ptr < end; j++)
+				ptr += sprintf(ptr, "%02x ", msg->buf[j]);
```
Yes. I need to figure out how to make checkpatch tell me

sjg20 replied:
done

sjg20 commented on drivers/i2c/i2c_common.c:
```diff
@@ -0,0 +1,37 @@
+			*data = '\0';
+		} else {
+			for (j = 0, ptr = data; j < msg->len && ptr < end; j++)
+				ptr += sprintf(ptr, "%02x ", msg->buf[j]);
```
No I didn't notice it.

I tried it and I see this:

I: Start program. Master: I2C_1, address: 0x54, off=2
D: eeprom: transfer I2C_1:
D: I2C msg: in, addr=54
D: I2C msg: 
D: 02                      |.       
D: I2C msg: 
D: 00 01 02 03 04 05 06 07 |........
D: 08 09 0a 0b 0c 0d 0e 0f |........
D: 10 11                   |..      
D: eeprom: transfer I2C_1:
D: I2C msg: in, addr=54
D: I2C msg: 
D: 02                      |.       
D: I2C msg: 
D: ff ff ff ff ff ff ff ff |........
D: ff ff ff ff ff ff ff ff |........
D: ff ff                   |..      

with no indication of read/write or the length. That function doesn't seem to support printf() strings (in fact it requires a persistent literal string). Also it always seems to print a newliine so it isn't possible to add things to the line.

So I think what I have is a lot nicer for the user:

I: Start program. Master: I2C_1, address: 0x54, off=2
D: eeprom: transfer I2C_1:
D: I2C msg: in, addr=54
D:    W len=01: 02 
D:    W len=12: 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 
D: eeprom: transfer I2C_1:
D: I2C msg: in, addr=54
D:    W len=01: 02 
D:    R len=12:

jfischer-phytec-iot replied:
I think the output of LOG_HEXDUMP_DBG is passable. Taking security and maintainability into account, one should not rewrite something the OS has a function for even if it does not look great.

sjg20 replied:
Pity the poor user.

Done

jfischer-phytec-iot replied:
I guess functions documentation should be updated too, https://github.com/zephyrproject-rtos/zephyr/pull/26003/commits/eb0522f904b97969441d3134b0765627a00bb410#diff-740314711ab1bb7517bee9c51dd569afR634

sjg20 replied:
Done

sjg20 commented on drivers/i2c/slave/eeprom_slave.c:
```diff
@@ -68,6 +68,34 @@ int eeprom_slave_read(struct device *dev, uint8_t *eeprom_data,
+	i2c_dump_msgs("in", msgs, num_msgs, addr);
+	if (num_msgs < 2 || (msg->flags & I2C_MSG_READ) || !msg->len) {
+		return 0;
+	}
```
done

sjg20 commented on drivers/i2c/i2c_common.c:
```diff
@@ -0,0 +1,37 @@
+			*data = '\0';
+		} else {
+			for (j = 0, ptr = data; j < msg->len && ptr < end; j++)
+				ptr += sprintf(ptr, "%02x ", msg->buf[j]);
```
done

sjg20 commented on include/drivers/i2c.h:
```diff
@@ -129,6 +129,27 @@ struct i2c_msg {
+ * Check your I2C driver to see which one it implements. If your
+ * simulator needs to be used by multiple I2C drivers, then you may
+ * need to implement both!
+ *
```
It does not require existing code to be modified. This is adding a new, simpler call to do a similar this. The STM32 supports 5 separate callbacks which have the same effect but better match how that driver works.

So we could modify STM32 to do things this way, but I'm not sure if there is any benefit. Also other drivers may split things out like STM32?

pabigot replied:
Are you referring to the STM32 HAL, or something in Zephyr?  And to an STM32 I2C simulator, or an STM32 I2C driver?  It looks like this could be used in a driver as an alternative to implementing the finer-grained ones.

AFAICT right now STM32 is the only target that supports the "slave" API, though I think some additional ones might be submitted.  I don't see anything like this in any I2C "slave" driver.

I can say that the original Zephyr I2C model of describing a multi-part transaction with a scatter/gather iovec-like API has caused significant pain, because not all I2C peripherals can support fragmenting bus transactions that way, and the API was not defined with precision so there are cross-driver differences in behavior and flag interpretation.  So the thought of adding another instance of it concerns me.

Though I concede that if presence/absence of an alternative implementation isn't visible to the user it wouldn't be an API change.  But then I'm not sure how anybody would need to implement multiple ones in a simulator.

Maybe I need a better understanding of exactly what you're aiming to do.  I thought it was just to run the existing I2C API tests on native POSIX in addition to real hardware.

sjg20 replied:
Yes that's right. But as I mentioned the current API for the slave is too fine-grained and seems to be tied in with STM32. I'd like a simple function like the one I added.

sjg20 replied:
> Maybe I need a better understanding of exactly what you're aiming to do

My goal is to have a way to write simulators for all the I2C peripherals, not just EEPROM. So for example we could have simulator (or fakes) for battery, accelerometer, lid sensor, temperature sensor, etc.

pabigot replied:
> My goal is to have a way to write simulators for all the I2C peripherals, not just EEPROM. So for example we could have simulator (or fakes) for battery, accelerometer, lid sensor, temperature sensor, etc.

OK.  So say we have a simulator for a Bosche BME280.  That simulator would presumably be something that hooks into an I2C-like controller simulator that recognizes specific device addresses and passes bus transactions to the corresponding device simulator to fill in response data.  So we'd be testing the real `drivers/sensor/bme280` but on native-posix with a faked I2C controller that's the bus parent of the faked sensor.  I assume `zephyr,i2c-sim` is intended to be that faked controller.

Is that use case in line with what you had in mind?

Maybe the core problem is that this test is intended to verify the I2C "slave" API, which is only implemented on one platform (STM32), and this test relies on a simulated device that in reality wouldn't provide an I2C "slave" interface.  A real EEPROM would be accessed through an I2C controller acting in "master" mode.  The test here checks the I2C "slave" mode implementation on STM32 only because there's a real STM32 I2C controller configured to that mode that uses a hack to inject data in response to driver-specific calls.

I don't think this is a good candidate for demonstrating the value of an I2C simulator.  Something like a BME280 would be better: if `samples/sensor/bme280` would run on native_posix with a BME280 simulator that could be somewhat useful, though I'd still go to real hardware.

Or, sticking with EEPROM, something that pretends to be an AT24C32 and passes the tests for `drivers/eeprom`.  Still a simulated controller on a simulated device, but I guess the production code for the eeprom driver would be tested (within the accuracy limits of the simulation of the device wrt timing and other edge conditions).

sjg20 commented on tests/drivers/i2c/i2c_slave_api/boards/stm32f072b_disco.overlay:
```diff
@@ -1,4 +1,8 @@
-/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright 2020 Google LLC
```
Just a mistake as I was testing it on that board. Done

sjg20 commented on drivers/i2c/i2c_common.c:
```diff
@@ -0,0 +1,37 @@
+		if (msg->flags & I2C_MSG_READ) {
+			*data = '\0';
+		} else {
+			for (j = 0, ptr = data; j < msg->len && ptr < end; j++)
```
I can change i but not j since there is another assignment. Done

pabigot committed:
tests: i2c: more test cleanup

Make the expected data buffers immutable to ensure that the test isn't
modifying them.

Improve clarity by using defines for the two devicetree nodes rather
than repeating the DT_INST() retrieval.

Clean up the naming and diagnostics to more clearly associate the
device label, I2C bus, and I2C address with either EEPROM instance 0
or 1.  Replace the base 1 numbering in some diagnostics.

Then fix all the cases where the test was trying to use the I2C
address for EP1 on the I2C bus for EP0 and vice versa.

Use the technique from portability/cmsis_rtos_v1 to compensate for
systems where times are being measured with different clocks.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Regarding https://github.com/zephyrproject-rtos/zephyr/pull/26003#discussion_r458159336:

I've looked more closely.  The original test was horribly wrong, and probably didn't even pass on the STM32.  I had provided fixes for the key problems in https://github.com/pabigot/zephyr/commits/pr/26003, but it doesn't look like I was clear enough that you needed to incorporate them.

I've again reworked it to fix the problems, this time in https://github.com/pabigot/zephyr/commits/pr/26003b.  This branch includes #27029 which removes the potential of the comparison buffer being mutated.

The fundamental problem is that the original test was crossing the I2C bus for EP0 with the I2C register for EP1 and vice versa.  Unfortunately the comparison data was paired with the (wrong) register rather than the (right) bus.  Which is why it looked like things were getting crossed.

Please do look at that branch, and replace the commit you added to make it work with dcb869a44e1972023ea3db8b1 at the head of https://github.com/pabigot/zephyr/commits/pr/26003b.  I believe you'll find that this works.

As an aside; that this worked at all suggests the EEPROM simulator is broken: it should have refused to accept an I2C operation for an EEPROM that wasn't using the correct associated device.

pabigot commented:
OK, we're getting closer.  You can rebase this on master as #27029 has been merged.

I've got the original master version of the test running on the nucleo_l476rg, where it passes.  With the version in this PR it fails.  I believe this has to do with the swapped device/address issues in the original test.  I'm looking into what the "eeprom_slave" driver is doing.

pabigot committed:
tests: i2c: add nucleo_l476rg as supported test platform

Add the EEPROM devices to the buses and document which pins must be
shorted to run the test.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Things are clearer now, but wow, this is messy.

First, I was wrong when I proposed changing the test so that the bus for EP0 was used for the I2C-based readback validation from the address for EP0.  The way this test works is with a loop-back where there are two controllers that share the SDA and SCL signals.  So when you're trying to read from EP0, you must issue the request using the controller for EP1 as the leader (master) because the controller for EP0 has to act as the follower (slave) to feed back the data.

I've replaced the content of the "tests: i2c: more test cleanup" commit to correct this: a comment has been expanded, the addr variable names got shortened, and the bus used in the I2C tests is always the one from the EEPROM that's not being checked. 
 I've also added a commit that adds support for an STM32 board I have.

I've verified that the board works correctly on real hardware with the test as it was before this PR, and with the modified PR in https://github.com/pabigot/zephyr/commits/pr/26003c.  You can take that branch as-is (it's rebased), or just drop the top commit from yours and add the top two commits dff0aa5a225092db85cc502f39c3718a7b50d923 and 88771b4ba3b18ec1e2d6373174f41ad9284bb13d from mine (and then rebase to get the #27029).

Now the problem: The test doesn't work on native posix.  That's because the simulator is incorrectly implemented.

The data in the `eeprom_slave` instance is associated with the I2C device that controls the bus it's attached to.  Thus EP0 is a device at addr_0 on i2c_0.  i2c_0 is the bus leader (master) for access to the device.

For the test to work the two controllers share the I2C signals, but you can't access EP0 as a follower (slave) device on i2c_0: you have to access it at addr_0 on the bus using i2c_1 as the leader (master).  i2c_0 then responds to those requests as a follower (slave) device.

In the test checking EP0 reads back correctly the bus transaction is controlled using i2c_1.  But the data that must be returned is associated with a device that is registered under i2c_0.

In real hardware the hand-off occurs because signals are shorted so a bus read issued on i2c_1 is fulfilled by a write from i2c_0.

The simulator doesn't know anything about any other devices, so it's returning the data from the device it knows about, which is the wrong EEPROM.  I.e. you're using the `slave_cfg` data attached to i2c_1 to provide the data, when you need to be using the `slave_cfg` data attached to i2c_0.

To make the simulator work you're going to need some way to indicate that i2c_0 and i2c_1 share bus signals.  Then when you're given a leader operation (write or read) to perform on one controller you need to search for a linked controller that has a registered follower device at the given address, and issue the complementary operation (read or write) on the linked controller.  Just like real hardware would do.

Making this work is not trivial.  An approach I think is feasible is to add a new devicetree compatible that identifies a set of `zephyr,sim-i2c` devices that share control signals.  A devicetree node for this would look like:

```
shared_i2c {
	compatible = "zephyr,sim-i2c-link";
	members = <&i2c1, &i2c2>;
};
```

Then the `i2c_sim.c` driver will have to populate an array recording the linked controllers, and each controller will have to have a pointer to that array.  (I think it's OK to assume there's only one set of linked controllers, i.e. only one instance of `zephyr,sim-i2c-link`, so you can just have a local array in the driver module.)

Then you'll need to implement the transfer operations to do a hand-off to the other controller as described above.  It may be easier to understand by implementing the finer-grained read and write request and completion functions, where the hand-off is more easily tracked.

I recognize this is probably badly explained, but I hope I've gotten the general idea across.  Drawing an object relationship diagram of the shared bus and the two EEPROM devices accessed as either leader (master) or follower (slave) and which bus device knows where the simulated data is stored might help if it's not clear why the current implementation is wrong.

Please let me know if you have questions.  If necessary we can schedule a telecon to work through it.

sjg20 commented:
Hi Peter,

Thanks for the details. I had not picked up on the fact that the current tests used a loop-back. Is that a hardware loopback (physically connecting pins) or does the chip support this?

In any case, for posix, I don't want to have two I2C buses connected together. I don't see any point in doing that. Instead, I want to model the I2C peripherals connected to the bus. So I can set up the peripheral and check its behaviour over I2C.

I designed a similar setup in U-Boot and it works well for posix tests (U-Boot calls this 'sandbox'). I'm happy to talk about all this if it suits you.

- Simon

pabigot commented:
I'm not sure how the test was supposed to work.  There's this `I2C_VIRTUAL` feature that's default-enabled for the test, but disabled in every actual board configuration supported by the test.  The test works on real hardware only when SDA and SCL for two hardware I2C controllers are shorted; this is regardless of whether `CONFIG_I2C_VIRTUAL=y`.  Swapping the controller devices as you needed to do for the simulator doesn't work with virtual either.

A cursory look suggests what i2c_virtual does is what I was saying needed to be done: provide a way that an operation driven by a leader-mode I2C controller can be associated with something that reacts like an I2C follower.  Perhaps @superna9999 can provide more insight on what that was intended to do and how it was to be used when introduced in 94d3156540d27392236044eebcb4c3a6e6065185.

In any case, if the simulator implemented here requires the test code be changed to pass wrong I2C controller in order to access the simulated subordinate device's data and make the test pass, it's not clear to me how this helps generally.

sjg20 commented:
Gosh this really is a bit of a mess. If we don't think the existing code works, should we delete it?

I've pushed this again with what I think is what you intended. Github doesn't seem to have support for comparing patch revisions though, so it is hard for me to know. If there is something wrong, a fix-up patch on top would be best I think.

Anyway I hope you understand what I am trying to do. Yes there are two I2C buses, but on posix they are independent and not tied together, or at least that is what I intend. I saw the existing 'slave' interface and thought I should try to use it, but perhaps that was a mistake?

I'm really looking forward to making some progress here as so much is blocked on this. This pull request is the most basic infrastructure to make things work, with just a single call.

pabigot commented:
I've pushed a fixup commit to https://github.com/pabigot/zephyr/commits/pr/26003d that adds the documentation of what's going on and fixes the parameters so the test passes on hardware.

I do think this is not the best starting point.  I'll reach out to you via slack to coordinate a plan.

sjg20 commented:
Thanks Peter for the valuable help and guidance,

I have updated the pull request to drop the posix implementation for now. Hopefully the remaining commits are useful to move things forward.

I'll do a separate pull request with a proposal 'v2' implementation as we discussed, once I get to it.

Regards,
Simon

cfriedt commented on tests/drivers/i2c/i2c_slave_api/boards/nucleo_l476rg.overlay:
```diff
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: Apache-2.0 */
```
Copyright statement is required

pabigot commented:
The QA check failure is in one of my commits.

This PR no longer has the material described in its summary.  I would like to see the remainder split into two PRs: one for the I2C common module, which I'd prefer to leave to @sjg20 to push forward, and one for the cleanup of the test case and its documentation, which needs some significant revisions since we now understand what it does better and the documentation doesn't really address how to use it properly.

For the latter the changes I'd like to see are significant, and I won't know exactly what they are until I've made and verified them.

@sjg20 There are several ways forward per the [modifying contributions](https://docs.zephyrproject.org/latest/contribute/modifying_contributions.html#modifying-contributions).  One is that I and others continue to ask you to make changes to these commits, with your branch remaining its home.  Another is that you close this PR (which no longer does what it was originally described to do).  You can open a PR for the I2C common module, and I'll open a PR for the test case cleanup, which would incorporate material from your commits as well as mine.  I'd ask you to be a reviewer for the latter.

Let me know how you want to proceed.

cfriedt commented:
Uh-oh - I didn't realize that this PR ran off the rails. I'm actively looking to use / create an i2c simulator for automated testing (and also a spi simulator). If I can be of any assistance, please let me know.

ElectricalPaul reviewed (commented):
So this test now only works on native_posix, and no longer will run on hardware?

pabigot commented:
> So this test now only works on native_posix, and no longer will run on hardware?

No.  It was originally modified to *also* work on native_posix using a simulator; that didn't end up working, so when the dust settles the test will (again) work on real hardware.  It'll also be better documented so people can figure out how to make it work on real hardware.

sjg20 commented:
OK I think it is best to do option 2. I've opened a new pull request here:

https://github.com/zephyrproject-rtos/zephyr/pull/27270

I'll try to keep an eye out for your one. In the meantime I'll work on a new i2c simulator as we discussed

@cfriedt hoping to have something by end of Monday
